README.md

```diff
@@ -1,277 +1,899 @@
+4) If "Yes", this is the corresponding output.
+
+It then compares the value of that output to the input to determine the swap's asset flux. Since the
+validator first checks for the beacon, each execution is dedicated for a specific trading pair. Any
+other outputs are ignored in this specific execution. This logic works because a script is executed
+once for every UTxO spent from the address. If input 1 is for beacon XYZ and input 2 is for beacon
+ABC, the first execution can be dedicated to beacon XYZ and the second execution can be dedicated to
+ABC. The net transaction will only succeed if all executions succeed. This behavior allows cheaply
+composing swaps of different trading pairs that are located at the same address. In other words, the
+logic takes advantage of the redundant executions.
+
+At a low-level, for a swap execution to be successfull, all of the following must be true:
+
+1. The input must contain the beacon for that trading pair - the required beacon is gotten from the
+   datum for that UTxO. If the beacon is present, that means the datum can be trusted.
+2. There must be an output to this address with the proper value and inline `SwapDatum`:
+    - Must contain exactly the same value as the input except for the ask asset and offer asset.
+    - All fields the same as the input datum except the `prevInput` field must be `Just(input_ref)`
+      where `input_ref` is the corresponding input's output reference.
+3. Offered asset taken * price <= asked asset given.
+4. Only the offered asset can leave and only the ask asset can be deposited. ADA can always be
+deposited in case the minUTxOValue increased.
+
+Requirement 1 guarantees that all invalid UTxOs (those missing beacons) belong to the address owner
+and that swap inputs have a valid price: denominator > 0 and price > 0.
+
+Requirements 2 & 4 guarantee that beacons from other trading pairs cannot be combined into one output
+UTxO. This has two beneficial consequences:
+
+1) The swap logic is very simple.
+2) All swap UTxOs are as small as possible which maximizes how many swaps can be composed in a
+single Tx as well as how easily the beacons can be queried.
+
+Custom error messages are included to help troubleshoot why a swap failed.
+
+
+## Benchmarks and Fee Estimations (YMMV)
+
+The protocol is capable of handling 11-14 swaps in a single transaction, regardless of the
+composition of one-way and two-way swaps in the transaction.
+
+**No CIPs or hard-forks are needed. This protocol works on the Cardano blockchain, as is.**
+
+Full benchmarking details can be found in the [Benchmarks](./Benchmarks/) folder.
 
 
-### Cardano-Swaps Lifecycle
-Below is an in-depth look at Cardano-Swaps logic.
+## Features Discussion
 
-#### Preparing a Swap Address
-Swap "writers" must first prepare a swap address by using the `MintBeacon` Redeemer in the following manner:
+### Full Delegation Control
```

Comment:
How's this section that would go before the delegation control section:

> ### Full Custody
> Since users get their own DEX addresses which use their own staking credentials for spending authorization, users maintain full custody of their assets at all times while using the DEX. *Not your keys, not your crypto.*

Reply:
Excellent: exactly what I had in mind 🤩

Reply:
I agree.  This feature is critical to the proper decentralised control of Cardano through the staking mechanism and voting in governance actions.  If users need to give up voting + staking control over their assets in order to use them in swaps or as collateral, then this severely impacts Cardano's resilience.